src/nemos/solvers.py

```diff
@@ -0,0 +1,725 @@
+            df_xs = self.loss_gradient(init_params, *args)
+
+        state = SVRGState(
+            iter_num=0,
+            key=self.key if self.key is not None else random.key(123),
+            error=jnp.inf,
+            stepsize=self.stepsize,
+            xs=init_params,
+            df_xs=df_xs,
+        )
+        return state
+
+    @partial(jit, static_argnums=(0,))
+    def _xk_update_step(
+        self,
+        xk: Pytree,
+        xs: Pytree,
+        df_xs: Pytree,
+        stepsize: float,
+        prox_lambda: Union[float, None],
+        *args,
+    ) -> Pytree:
+        """
+        Body of the inner loop of Prox-SVRG that takes a step.
+
+        Parameters
+        ----------
+        xk :
+            Current parameters.
+        xs :
+            Anchor point.
+        df_xs :
+            Full gradient at the anchor point.
+        stepsize :
+            Step size.
+        prox_lambda :
+            Hyperparameters to `prox`, most commonly regularization strength.
+        args:
+            Positional arguments passed to loss function `fun` and its gradient (e.g. `fun(current_params, *args)`),
+            most likely input and output data.
+            They are expected to be Pytrees with arrays or FeaturePytree as their leaves, with all of their
+            leaves having the same sized first dimension (corresponding to the number of data points).
+            For GLMs these are:
+                X : DESIGN_INPUT_TYPE
+                    Input data.
+                y : jnp.ndarray
+                    Output data.
+
+        Returns
+        -------
+        next_xk :
+            Parameter values after applying the update.
+        """
+        # gradient on batch_{i_k} evaluated at the current parameters (xk)
+        grad_of_fik_at_xk = self.loss_gradient(xk, *args)
+        # gradient on batch_{i_k} evaluated at the anchor point (xs)
+        grad_of_fik_at_xs = self.loss_gradient(xs, *args)
+
+        # SVRG gradient estimate
+        gk = jax.tree_util.tree_map(
+            lambda a, b, c: a - b + c, grad_of_fik_at_xk, grad_of_fik_at_xs, df_xs
+        )
+
+        # x_{k+1} = x_{k} - stepsize * g_{k}
+        next_xk = tree_add_scalar_mul(xk, -stepsize, gk)
+
+        # apply the proximal operator
+        next_xk = self.proximal_operator(next_xk, prox_lambda, scaling=stepsize)
+
+        return next_xk
+
+    @partial(jit, static_argnums=(0,))
+    def update(
+        self,
+        current_params: Pytree,
+        state: SVRGState,
+        prox_lambda: Union[float, None],
+        *args,
+    ) -> OptStep:
+        """
+        Perform a single parameter update on the passed data (no random sampling or loops)
+        and increment `state.iter_num`.
+
+        Please note that this is called by `GLM.update`, but repeated calls to `GLM.update`
+        on mini-batches passed to it will not result in running the full (Prox-)SVRG,
+        and parts of the algorithm will have to be implemented outside.
```

Comment:
- currently, this gets called by `GLM.update`, but it will get called by any model update step, right? I would say something like "This gets called by `BaseRegressor.solver_update` (e.g., as called by `GLM.update`)..."
- For "repeated calls ... will not result in running the full SVRG": in order to run the full SVRG, they should call `.run`, right? We should say that instead of making it sound like they need to implement it themselves

Reply:
- Re `BaseRegressor._solver_update`: Yes, I'll change it.
- Re `.run`: `.run` and `.fit` assumes you have the full data in memory and can index into it. If you don't, you have to rely on `.update` and you do have to implement some parts of the SVRG algorithm yourself. This is in contrast with SGD (shown in the batching example), where you just have to call `.update` on each new minibatch. We have tests that the full SVRG algorithm calling `.update`, but we still need to make a tutorial in the docs. In the future this would be hidden from the user with an alternative interface for batched fitting. These might be relevant:
https://github.com/flatironinstitute/nemos/issues/194
https://github.com/flatironinstitute/nemos/issues/203

Reply:
I addressed the first point. Let me know how you like the new description.

Reply:
ah that's right, I forgot about that. as long as we remember to show it in the tutorial, that's fine.